dataline-workers/src/main/java/io/dataline/workers/DefaultSyncWorker.java

```diff
@@ -72,15 +72,17 @@ public DefaultSyncWorker(String tapDockerImage, String targetDockerImage) {
 
     final SingerMessageTracker singerMessageTracker =
         new SingerMessageTracker(syncInput.getStandardSync().getConnectionId());
-    try (Stream<SingerMessage> tap =
-        new SingerTapFactory(tapDockerImage).create(tapConfig, workspacePath)) {
-      tapCloser = tap::close;
 
-      final Stream<SingerMessage> peakedTapStream = tap.peek(singerMessageTracker);
-      try (Target<SingerMessage> target = new SingerTarget(targetDockerImage)) {
-        targetCloser = getTargetCloser(target);
+    final SingerTapFactory singerTapFactory = new SingerTapFactory(tapDockerImage);
+    final SingerTargetFactory singerTargetFactory = new SingerTargetFactory(targetDockerImage);
+
+    try (Stream<SingerMessage> tap = singerTapFactory.create(tapConfig, workspacePath)) {
+      tapCloser = tap::close;
 
-        target.consume(peakedTapStream, targetConfig, workspacePath);
+      try (CloseableConsumer<SingerMessage> consumer =
```

Comment:
you can create both resources in the same try:
`try (tap = xxx; consumer = yyy)`